qualtran/_infra/gate_with_registers.py

```diff
@@ -549,3 +546,51 @@ def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs) -> cirq.Circ
 
         wire_symbols[0] = self.__class__.__name__
         return cirq.CircuitDiagramInfo(wire_symbols=wire_symbols)
+
+
+def get_ctrl_system_for_single_qubit_controlled(
+    bloq: GateWithRegisters, ctrl_spec: Optional['CtrlSpec'] = None
+) -> Tuple['Bloq', 'AddControlledT']:
+    """Wrapper to correctly wire a custom implementation of a single-qubit controlled version of a gate.
+
+    For any gate that supports a single-qubit controlled version
+    with an API that has `control_val` and `control_registers`.
+    Add the `get_ctrl_system` implementation as shown below:
+
+    .. code:: python
+
+        class MyGate(GateWithRegisters):
+            control_val: Optional[int]
+
+            @property
+            def control_registers() -> Tuple[Register, ...]:
+                ...
+
+            def get_ctrl_system(
+                self, ctrl_spec: Optional['CtrlSpec'] = None
+            ) -> Tuple['Bloq', 'AddControlledT']:
+                from qualtran._infra.gate_with_registers import get_ctrl_system_for_single_qubit_controlled
+
+                return get_ctrl_system_for_single_qubit_controlled(self, ctrl_spec)
+    """
+    if ctrl_spec is None:
+        ctrl_spec = CtrlSpec()
+
+    assert hasattr(bloq, 'control_val')
+    assert hasattr(bloq, 'control_registers')
+
+    if bloq.control_val is None and ctrl_spec.shapes in [((),), ((1,),)]:
+        cbloq = attrs.evolve(bloq, control_val=int(ctrl_spec.cvs[0].item()))
+        (ctrl_reg,) = cbloq.control_registers
+
+        def adder(
+            bb: 'BloqBuilder', ctrl_soqs: Sequence['SoquetT'], in_soqs: dict[str, 'SoquetT']
+        ) -> tuple[Iterable['SoquetT'], Iterable['SoquetT']]:
+            soqs = {ctrl_reg.name: ctrl_soqs[0]} | in_soqs
+            soqs = bb.add_d(cbloq, **soqs)
+            ctrl_soqs = [soqs.pop(ctrl_reg.name)]
+            return ctrl_soqs, soqs.values()
+
+        return cbloq, adder
+
+    raise NotImplementedError(f'Cannot create a controlled version of {bloq} with {ctrl_spec=}.')
```

Comment:
for this to properly implement the `get_ctrl_system` method, it should delegate to the base class method if no specific construction is available. (This base method will wrap the bloq in `Controlled` metabloq).

Reply:
fixed